Remove Job from Info API

HuKeping opened the pull request:
a part of issue #12151

Signed-off-by: Hu Keping hukeping@huawei.com

duglin commented on daemon/info.go:
```diff
@@ -15,7 +15,7 @@ import (
 )
 
-func (daemon *Daemon) CmdInfo(job *engine.Job) error {
+func (daemon *Daemon) Info() (*engine.Env, error) {
```
Even if the result is just an array of strings, let's either create a new type in api/types or just use `[]string` instead of engine.Env because I'm hoping engine.Env goes away, eventually.

duglin replied:
Is there any reason we're not typing all of the results here so its not just a list of strings being returned to the client?

LK4D4 replied:
Yup, +1. Env is superbad.

HuKeping replied:
get it

HuKeping committed:
Remove Job from Info API

Two main things
- Create a real struct Info for all of the data with the proper types
- Add test for REST API get info

Signed-off-by: Hu Keping <hukeping@huawei.com>

LK4D4 commented on api/server/server.go:
```diff
@@ -321,8 +321,11 @@ func getImagesViz(eng *engine.Engine, version version.Version, w http.ResponseWr
 	w.Header().Set("Content-Type", "application/json")
-	eng.ServeHTTP(w, r)
-	return nil
+	if env, err := getDaemon(eng).Info(); err != nil {
```
It is better to use simple form without else. Also it is recommended by google:
https://code.google.com/p/go-wiki/wiki/CodeReviewComments#Indent_Error_Flow

HuKeping replied:
I'll update this

jfrazelle commented on api/server/server_unit_test.go:
```diff
@@ -64,33 +64,6 @@ func TestGetVersion(t *testing.T) {
 	}
 }
 
-func TestGetInfo(t *testing.T) {
```
we should also reimplement the test in integration-cli

HuKeping replied:
ok, test will come soon

HuKeping commented:
Updated-add a structure Info, is this change good to you? ping  @LK4D4 @duglin

LK4D4 commented on api/server/server.go:
```diff
@@ -340,8 +340,13 @@ func getImagesViz(eng *engine.Engine, version version.Version, w http.ResponseWr
+		return err
+	}
+
+	return writeJSON(w, http.StatusOK, *info)
```
Just pass `info`, should work.

LK4D4 commented:
Ok for me. What about test? Maybe there is already some in `integration-cli`.

HuKeping commented:
I am plan to add a new file integration-cli/docker_api_info_test.go to do the test. It's on the way.

HuKeping commented:
Test ready. @LK4D4  @jfrazelle 

Since the CLI command `docker info` has been checked by the docker_cli_info_test.go, I think there is only lack of checking for REST API `GET /info`. I put it into a new file docker_api_info_test.go.

duglin commented:
Please squash your commits down to one.
Overall this looks good.
We need to also modify the CLI code to use a structure too. Would you like to do that one as well? If so, we could do it in this PR at the same time if you're up for it.  If not we can do it in a separate PR.  Just let us know.

LK4D4 commented on integration-cli/docker_api_info_test.go:
```diff
@@ -0,0 +1,15 @@
+func TestInfoApi(t *testing.T) {
+	endpoint := "/info"
+
+	if _, err := sockRequest("GET", endpoint, nil); err != nil {
```
`sockRequest` changed signature in master. Now you can get `StatusCode` and check it. Also maybe makes sense to unmarshall body to some temp structure for checking that important fields is in there.

HuKeping replied:
ok

HuKeping commented:
@duglin I'll do it.

HuKeping commented:
update.
- make change on CLI command `docker info`
- upgrade test

HuKeping commented:
I'm trying to fix this CI error:
daemon/execdriver/native/exec.go:13:2: C source files not allowed when not using cgo: nsexec.c

HuKeping commented:
ping @duglin  @LK4D4

HuKeping commented:
rebase, ping

cpuguy83 commented on api/client/info.go:
```diff
@@ -19,127 +19,76 @@ func (cli *DockerCli) CmdInfo(args ...string) error {
-		logrus.Errorf("Error reading remote info: %s", err)
+	info := &types.Info{}
+	if err := json.NewDecoder(rdr).Decode(info); err != nil {
+		logrus.Errorf("Error reading remote info: %v", err)
```
I don't think we need this error message here since we're returning error below.
Maybe update the error with this extra info, though?

cpuguy83 commented:
:( needs rebase.

HuKeping commented:
@cpuguy83 rebased

HuKeping commented:
ping @duglin @LK4D4

LK4D4 commented on integration-cli/docker_api_info_test.go:
```diff
@@ -0,0 +1,38 @@
+	"testing"
+)
+
+func TestInfoApi(t *testing.T) {
```
Can we have simple test for `cli` too? Because it was changed. Or maybe there is already tests?

HuKeping replied:
We have it under `integration-cli/docker_cli_info_test.go`

LK4D4 commented:
LGTM

cpuguy83 commented:
LGTM
